### DISCUSSION TITLE
New Game Mode: Minesweeper

### DISCUSSION THREAD

**Discussion by robbles** [BattlesnakeOfficial/feedback#83] [General]
TL;DR: Some foods are secretly "mines" that instantly kill your snake. Eating a "safe" food gives you health proportional to the number of adjacent mines.

Movement: No change
Board layout:  No change
Food: All food is spawned at start, and there's a lot more of it. Every mine food is surrounded by safe foods. Eating a mine food instantly kills the snake.
Health: Health increases by 1 when consuming a safe food adjacent to 1 mine, by 2 if it's adjacent to 2 mines, and so on. Health is not capped at 100, so that the change in health can always be measured between turns.

I think this one could be interesting, because (I think) it forces a small amount of game state to be kept between turns by a snake. Otherwise you can't tell how much health you picked up.

![minesweeper_small](https://user-images.githubusercontent.com/92927/125716977-bcaef829-ad11-4040-8f8e-46bbef0304ed.png)

**Comment by altersaddle** [comment]
This game mode could also be implemented using hazards as the 'hint' squares, since:

- Hazard squares may appear at any point 
- Snakes may enter hazard squares freely

However, the board API currently presents hazards as a list of coordinate pairs.  An API change would be required to add the necessary information, but I think this will have the intended effect as in minesweeper - when you click an empty square, the game explores the empty space around that square and reveals clue tiles to fill the space.

Bomb squares could still be represented with food, by tweaking server rules so that food results in -100 health debuff, and moving does not cost health.

There is a minor problem if using regular battlesnake move rules - there may be places to explore on the map that cannot be reached (dead ends).  If this is a single-player challenge mode, then perhaps we permit the snake to collide with its body to allow better exploration?  

What about diagonal travel, is that possible somehow?

**Reply by robbles** [reply]
I might be imagining a different scenario here, but I think unreachable squares could probably be addressed somewhat with slightly smarter grid/snake layout by the engine?
And possibly some of it is just part of the challenge :)

**Reply by robbles** [reply]
I like the idea of using hazards for this. Possibly they would even work out of the box by just "stacking" them? e.g. 2 hazards with the same coords is a 2 on the board, and so on.
